core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java

```diff
@@ -312,21 +334,26 @@ public void requestQuorumStateUpdate(Peer downloadPeer, StoredBlock requestBlock
         }
     }
 
-    public void processDiffMessage(Peer peer, QuorumRotationInfo qrinfo, boolean isLoadingBootStrap) {
-        processQuorumRotationInfo(peer, qrinfo, isLoadingBootStrap);
+    public void processDiffMessage(Peer peer, QuorumRotationInfo qrinfo, boolean isLoadingBootStrap, @Nullable SettableFuture<Boolean> opComplete) {
+        processQuorumRotationInfo(peer, qrinfo, isLoadingBootStrap, opComplete);
     }
 
-    public void processQuorumRotationInfo(@Nullable Peer peer, QuorumRotationInfo quorumRotationInfo, boolean isLoadingBootStrap) {
-        try {
-            quorumRotationState.processDiff(peer, quorumRotationInfo, headersChain, blockChain, isLoadingBootStrap);
-            processMasternodeList(quorumRotationInfo.getMnListDiffAtH());
-            setFormatVersion(BLS_SCHEME_FORMAT_VERSION);
-            unCache();
-            if (quorumRotationInfo.hasChanges() || quorumRotationState.getPendingBlocks().size() < MAX_CACHE_SIZE || saveOptions == SimplifiedMasternodeListManager.SaveOptions.SAVE_EVERY_BLOCK)
-                save();
-        } finally {
-            // TODO: do we need a finally?
-        }
+    public void processQuorumRotationInfo(@Nullable Peer peer, QuorumRotationInfo quorumRotationInfo, boolean isLoadingBootStrap, @Nullable SettableFuture<Boolean> opComplete) {
+
+        // process qrinfo asynchronously
+        threadPool.execute(new Runnable() {
+            @Override
+            public void run() {
+                quorumRotationState.processDiff(peer, quorumRotationInfo, headersChain, blockChain, isLoadingBootStrap);
+                processMasternodeList(quorumRotationInfo.getMnListDiffAtH());
+                setFormatVersion(BLS_SCHEME_FORMAT_VERSION);
+                unCache();
+                if (quorumRotationInfo.hasChanges() || quorumRotationState.getPendingBlocks().size() < MAX_CACHE_SIZE || saveOptions == SimplifiedMasternodeListManager.SaveOptions.SAVE_EVERY_BLOCK)
+                    save();
+                if (opComplete != null)
+                    opComplete.set(true);
+            }
+        });
     }
```

Comment:
originally, I thought this could be a run and forget thread (though we keep track of the `threadPool` to shut it down later.  For some tests, it is necessary to wait for this to finish execution.  Therefore a future argument was added in cases where the caller needs to know when  the operation has completed.